js-runtime/index.js

```diff
@@ -142,59 +166,469 @@ module.exports = {
   boolstr:  a => a.toString(),
 
   // Arithmetic opcodes
-  addi8:   (a, b) => a + b,
-  addi16:  (a, b) => a + b,
-  addi32:  (a, b) => a + b,
-  addi64:  (a, b) => a + b,
-  addf32:  (a, b) => a + b,
-  addf64:  (a, b) => a + b,
-
-  subi8:   (a, b) => a - b,
-  subi16:  (a, b) => a - b,
-  subi32:  (a, b) => a - b,
-  subi64:  (a, b) => a - b,
-  subf32:  (a, b) => a - b,
-  subf64:  (a, b) => a - b,
+  addi8:   (ra, rb) => {
+    if (!ra[0]) return ra
+    if (!rb[0]) return rb
+    const a = ra[1]
+    const b = rb[1]
+    if (a === 0 || b === 0) return [1, a + b]
+    if (a < 0 && b > 0) return [1, a + b]
+    if (a > 0 && b < 0) return [1, a + b]
+    if (a > 0 && b > 0) {
+      if (a > INT8MAX - b) return [0, 'overflow']
+      return [1, a + b]
+    }
```

Comment:
When you *might* have a case that could fail, do the math to check it, and return an error result if it would have failed.